### PR TITLE
Better checkboxes and radio inputs

### LINKED ISSUE #21: Better checkbox/radio styles
cf https://github.com/dobtco/screendoor-v2/issues/263

### PR DESCRIPTION
**Pull request by jrubenoff**
Closes #21

How are we handling migration?

### CODE REVIEW

File: spec/dummy/app/views/base.rb [hunk 1]
```diff
@@ -3,6 +3,7 @@ def content
     rawtext '<!doctype html>'
     html {
       head {
+        link href: '//maxcdn.bootstrapcdn.com/font-awesome/4.2.0/css/font-awesome.min.css', rel: 'stylesheet'
```
Comment (dobthound) [inline spec/dummy/app/views/base.rb]: Line is too long. [109/80]

File: spec/dummy/app/views/home/index.rb [hunk 2]
```diff
@@ -39,14 +39,26 @@ def main
     docs 'Forms' do
       form {
         input type: 'text'
-        label.checkbox {
+        label.control {
```
Comment (dobthound) [inline spec/dummy/app/views/home/index.rb]: Avoid using {...} for multi-line blocks.

File: spec/dummy/app/views/home/index.rb [hunk 3]
```diff
@@ -39,14 +39,26 @@ def main
...
           input type: 'checkbox'
-          text 'Foo'
+          span.control_styled
+          text 'Do you agree?'
+        }
+        label.control {
```
Comment (dobthound) [inline spec/dummy/app/views/home/index.rb]: Avoid using {...} for multi-line blocks.

File: spec/dummy/app/views/home/index.rb [hunk 4]
```diff
@@ -39,14 +39,26 @@ def main
...
+          input type: 'radio', name: 'foo'
+          span.control_styled
+          text 'Toggle this radio'
+        }
+        label.control {
```
Comment (dobthound) [inline spec/dummy/app/views/home/index.rb]: Avoid using {...} for multi-line blocks.

File: spec/dummy/app/views/home/index.rb [hunk 5]
```diff
@@ -39,14 +39,26 @@ def main
...
+          input type: 'radio', name: 'foo'
+          span.control_styled
+          text 'Or this radio'
         }
         div.input_sub {
-          label.checkbox {
+          label.control {
```
Comment (dobthound) [inline spec/dummy/app/views/home/index.rb]: Avoid using {...} for multi-line blocks.

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 6]
```diff
@@ -176,72 +172,158 @@ select.datetime {
...
+    display: inline-block;
+    opacity: 0;
   }
-  &.readonly,
-  &.disabled {
-    opacity: 0.5;
+}
+
+label.control {
+    display: block;
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Line should be indented 2 spaces, but was indented 4 spaces

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 7]
```diff
@@ -176,72 +172,158 @@ select.datetime {
...
+    font-weight: normal;
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Line should be indented 2 spaces, but was indented 4 spaces

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 8]
```diff
@@ -176,72 +172,158 @@ select.datetime {
...
+    font-size: 1rem;
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Line should be indented 2 spaces, but was indented 4 spaces

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 9]
```diff
@@ -176,72 +172,158 @@ select.datetime {
...
+    position: relative;
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Line should be indented 2 spaces, but was indented 4 spaces

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 10]
```diff
@@ -176,72 +172,158 @@ select.datetime {
...
+    line-height: 1.5rem;
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Line should be indented 2 spaces, but was indented 4 spaces

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 11]
```diff
@@ -176,72 +172,158 @@ select.datetime {
...
+    input {
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Line should be indented 2 spaces, but was indented 4 spaces

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 12]
```diff
@@ -176,72 +172,158 @@ select.datetime {
...
-    color: $linkColor;
+    display: inline-block;
+    opacity: 0;
   }
-  &.readonly,
-  &.disabled {
-    opacity: 0.5;
+}
+
+label.control {
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Merge rule `label.control` with rule on line 148<br>Avoid qualifying class selectors with an element.

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 13]
```diff
@@ -83,6 +83,20 @@ textarea {
...
+// Focus states
+
+input[type="text"],
+input[type="email"],
+input[type="search"],
+input[type="password"],
+input[type="url"],
+textarea {
+  &:focus {
+    border-color: $darkBlue;
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Name of variable `darkBlue` should be written in all lowercase letters with hyphens instead of underscores

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 14]
```diff
@@ -176,72 +172,158 @@ select.datetime {
...
+    font-size: 1rem;
+    position: relative;
+    line-height: 1.5rem;
+    input {
+      opacity: 0;
+      position: absolute;
+      top: 0;
+      left: 0;
+      &:focus + .control_styled {
+        border-color: $darkBlue;
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Name of variable `darkBlue` should be written in all lowercase letters with hyphens instead of underscores

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 15]
```diff
@@ -176,72 +172,158 @@ select.datetime {
...
+
+input[type="checkbox"] {
+  + .control_styled {
+    border-radius: 2px;
+    position: relative;
+    &:after {
+      position: absolute;
+      font-family: 'FontAwesome';
+      content: "\f00c";
+      color: $darkBlue;
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Name of variable `darkBlue` should be written in all lowercase letters with hyphens instead of underscores

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 16]
```diff
@@ -176,72 +172,158 @@ select.datetime {
...
+      line-height: 16px;
+      width: 16px;
+      text-align: center;
+      display: inline-block;
+      opacity: 0;
+      transform: scale(0.5) translateZ(0);
+      transition: opacity 120ms $easeInOutBack, transform 170ms $easeInOutBack;
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Name of variable `easeInOutBack` should be written in all lowercase letters with hyphens instead of underscores

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 17]
```diff
@@ -176,72 +172,158 @@ select.datetime {
...
-    height: 6px;
+  + .control_styled {
     border-radius: 50%;
-    position: absolute;
-    top: 4px;
-    left: 4px;
-    background-color: $linkColor;
+    &:after {
+      content: "";
+      background: $darkBlue;
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Name of variable `darkBlue` should be written in all lowercase letters with hyphens instead of underscores

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 18]
```diff
@@ -176,72 +172,158 @@ select.datetime {
...
+      width: 8px;
+      height: 8px;
+      border-radius: 50%;
+      position: absolute;
+      top: 50%;
+      left: 50%;
+      margin: -4px 0 0 -4px; // Center within the input
+      opacity: 0;
+      transform: scale(0.75) translateZ(0);
+      transition: opacity 120ms $easeInOutBack, transform 170ms $easeInOutBack;
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Name of variable `easeInOutBack` should be written in all lowercase letters with hyphens instead of underscores

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 19]
```diff
@@ -83,6 +83,20 @@ textarea {
   }
 }
 
+// Focus states
+
+input[type="text"],
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Avoid qualifying attribute selectors with an element.

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 20]
```diff
@@ -176,72 +172,158 @@ select.datetime {
...
+      }
+      &:checked + .control_styled:after {
+        opacity: 1;
+        transform: scale(1) translateZ(0);
+      }
+    }
+}
+
+
+input[type="checkbox"] {
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Avoid qualifying attribute selectors with an element.

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 21]
```diff
@@ -176,72 +172,158 @@ select.datetime {
...
-  -webkit-appearance: none !important;
-  -moz-appearance: none !important;
-  appearance: none !important;
-  border: 1px solid $gray;
-  border-radius: $radius;
-  background-color: #fff;
-  padding: 7px;
+// Checkbox and radio inputs
+
+.control_styled {
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Selector `control_styled` should be written in lowercase with hyphens

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 22]
```diff
@@ -176,72 +172,158 @@ select.datetime {
...
+    font-weight: normal;
+    font-size: 1rem;
+    position: relative;
+    line-height: 1.5rem;
+    input {
+      opacity: 0;
+      position: absolute;
+      top: 0;
+      left: 0;
+      &:focus + .control_styled {
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Selector `control_styled` should be written in lowercase with hyphens

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 23]
```diff
@@ -176,72 +172,158 @@ select.datetime {
...
+        border-color: $darkBlue;
+      }
+      &:hover + .control_styled {
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Selector `control_styled` should be written in lowercase with hyphens

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 24]
```diff
@@ -176,72 +172,158 @@ select.datetime {
...
+        background: #f9f9f9;
+      }
+      &:active + .control_styled {
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Selector `control_styled` should be written in lowercase with hyphens

**Comment by ajb** [comment]
> How are we handling migration?

Once this is merged, one of us (I'll volunteer) will open a branch in Screendoor called `update-dvl-core`, run `bundle update dvl-core`, and adjust the simple_form configuration (and any non-simple form markup).

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 25]
```diff
@@ -176,73 +172,111 @@ select.datetime {
...
+    display: inline-block;
+    opacity: 0;
   }
-  &:checked:after {
-    font-family: 'FontAwesome';
-    content: '\f00c';
-    font-size: 12px;
+}
+
+label.control {
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Merge rule `label.control` with rule on line 148<br>Avoid qualifying class selectors with an element.

File: vendor/assets/stylesheets/dvl/core/forms.scss (hunk 13 above)
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Name of variable `darkBlue` should be written in all lowercase letters with hyphens instead of underscores

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 26]
```diff
@@ -176,73 +172,111 @@ select.datetime {
...
+    left: 0;
+    &.readonly,
+    &.disabled {
+      + .control_styled {
+        opacity: 0.6;
+        background: #f5f5f5;
+      }
+    }
+    &:focus + .control_styled {
+      border-color: $darkBlue;
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Name of variable `darkBlue` should be written in all lowercase letters with hyphens instead of underscores

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 27]
```diff
@@ -176,73 +172,111 @@ select.datetime {
...
+      opacity: 0;
+      transform: scale(0.5) translateZ(0);
+      transition: opacity 120ms $easeInOutBack, transform 170ms $easeInOutBack;
+    }
   }
   &.readonly,
   &.disabled {
-    opacity: 0.5;
+    + .control_styled:after {
+      color: $darkBlue;
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Name of variable `darkBlue` should be written in all lowercase letters with hyphens instead of underscores

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 28]
```diff
@@ -176,73 +172,111 @@ select.datetime {
...
-  position: relative;
-  cursor: pointer;
-  &:active,
-  &:checked:active {
-    box-shadow: 0 1px 2px rgba(0,0,0,0.05), inset 0px 1px 3px rgba(0,0,0,0.1);
+  + .control_styled {
+    border-radius: 50%;
+    &:after {
+      content: "";
+      background: $darkBlue;
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Name of variable `darkBlue` should be written in all lowercase letters with hyphens instead of underscores

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 29]
```diff
@@ -176,73 +172,111 @@ select.datetime {
...
-    content: '';
-    width: 6px;
-    height: 6px;
-    border-radius: 50%;
-    position: absolute;
-    top: 4px;
-    left: 4px;
-    background-color: $linkColor;
+    + .control_styled:after {
+      background: $darkBlue;
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Name of variable `darkBlue` should be written in all lowercase letters with hyphens instead of underscores

File: vendor/assets/stylesheets/dvl/core/forms.scss (hunk 19 above)
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Avoid qualifying attribute selectors with an element.

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 30]
```diff
@@ -176,73 +172,111 @@ select.datetime {
...
+    }
+    &:checked + .control_styled:after {
+      opacity: 1;
+      transform: scale(1) translateZ(0);
+    }
+  }
+}
+
+
+input[type="checkbox"] {
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Avoid qualifying attribute selectors with an element.

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 31]
```diff
@@ -176,73 +172,111 @@ select.datetime {
...
-  -webkit-appearance: none !important;
-  -moz-appearance: none !important;
-  appearance: none !important;
-  border: 1px solid $gray;
-  border-radius: $radius;
-  background-color: #fff;
-  padding: 7px;
+// Checkbox and radio inputs
+
+.control_styled {
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Selector `control_styled` should be written in lowercase with hyphens

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 32]
```diff
@@ -176,73 +172,111 @@ select.datetime {
...
+    opacity: 0;
     position: absolute;
-    top: 1px;
-    left: 1px;
-    color: $linkColor;
+    top: 0;
+    left: 0;
+    &.readonly,
+    &.disabled {
+      + .control_styled {
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Selector `control_styled` should be written in lowercase with hyphens

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 33]
```diff
@@ -176,73 +172,111 @@ select.datetime {
...
+        background: #f5f5f5;
+      }
+    }
+    &:focus + .control_styled {
+      border-color: $darkBlue;
+    }
+    &:hover + .control_styled {
+      background: #f9f9f9;
+    }
+    &:active + .control_styled {
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Selector `control_styled` should be written in lowercase with hyphens

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 34]
```diff
@@ -176,73 +172,111 @@ select.datetime {
...
+      background: #f1f1f1;
+    }
+    &:checked + .control_styled:after {
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Selector `control_styled` should be written in lowercase with hyphens

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 35]
```diff
@@ -176,73 +172,111 @@ select.datetime {
...
+      opacity: 1;
+      transform: scale(1) translateZ(0);
+    }
+  }
+}
+
+
+input[type="checkbox"] {
+  + .control_styled {
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Selector `control_styled` should be written in lowercase with hyphens

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 36]
```diff
@@ -176,73 +172,111 @@ select.datetime {
...
+      display: inline-block;
+      opacity: 0;
+      transform: scale(0.5) translateZ(0);
+      transition: opacity 120ms $easeInOutBack, transform 170ms $easeInOutBack;
+    }
   }
   &.readonly,
   &.disabled {
-    opacity: 0.5;
+    + .control_styled:after {
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Selector `control_styled` should be written in lowercase with hyphens

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 37]
```diff
@@ -176,73 +172,111 @@ select.datetime {
...
-  border: 1px solid $gray;
-  border-radius: 50%;
-  padding: 7px;
-  display: inline-block;
-  position: relative;
-  cursor: pointer;
-  &:active,
-  &:checked:active {
-    box-shadow: 0 1px 2px rgba(0,0,0,0.05), inset 0px 1px 3px rgba(0,0,0,0.1);
+  + .control_styled {
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Selector `control_styled` should be written in lowercase with hyphens

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 38]
```diff
@@ -176,73 +172,111 @@ select.datetime {
...
-  &:checked:after {
-    content: '';
-    width: 6px;
-    height: 6px;
-    border-radius: 50%;
-    position: absolute;
-    top: 4px;
-    left: 4px;
-    background-color: $linkColor;
+    + .control_styled:after {
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Selector `control_styled` should be written in lowercase with hyphens

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 39]
```diff
@@ -176,73 +172,111 @@ select.datetime {
...
+}
+
+
+input[type="checkbox"] {
+  + .control_styled {
+    border-radius: 2px;
+    position: relative;
+    &:after {
+      position: absolute;
+      font-family: 'FontAwesome';
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Prefer double-quoted strings

**Comment by jrubenoff** [comment]
I made some fixes from Hound's comments and ignoring the rest. Please review + merge.

**Comment by ajb** [comment]
Here's the results from Browserstack's screenshots: http://www.browserstack.com/screenshots/d6cea63995346865860530b1d794f0beeb584bd7

tl;dr everything looks good except IE 8 + 7, which we don't support.

**Comment by ajb** [comment]
Working on the simple_form configuration for this stuff before merge.

File: lib/dvl/simple_form_config.rb [hunk 40]
```diff
@@ -0,0 +1,55 @@
+SimpleForm::Inputs::Base.default_options = {
+  # Add the 'control' class to radio/checkbox labels
+  item_label_class: 'control'
+}
+
+SimpleForm.setup do |config|
+  config.wrappers :horizontal, tag: 'div', class: 'grid gutter_none', error_class: 'error' do |b|
```
Comment (dobthound) [inline lib/dvl/simple_form_config.rb]: Line is too long. [97/80]

File: lib/dvl/simple_form_config.rb [hunk 41]
```diff
@@ -0,0 +1,55 @@
...
+    b.use :html5
+    b.use :placeholder
+    b.use(:judge) if defined?(Judge)
+
+    b.wrapper tag: 'div', class: 'item lap_three_columns horizontal_label_item' do |ba|
```
Comment (dobthound) [inline lib/dvl/simple_form_config.rb]: Line is too long. [87/80]

File: lib/dvl/simple_form_config.rb [hunk 42]
```diff
@@ -0,0 +1,55 @@
...
+    end
+
+    b.wrapper tag: 'div', class: 'item lap_nine_columns' do |ba|
+      ba.use :input
+      ba.use :error, wrap_with: { tag: 'span', class: 'error_message' }
+      ba.use :hint,  wrap_with: { tag: 'div', class: 'help-block' }
+    end
+  end
+
+  config.wrappers :vertical, tag: 'div', class: 'grid gutter_none', error_class: 'error' do |b|
```
Comment (dobthound) [inline lib/dvl/simple_form_config.rb]: Line is too long. [95/80]

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 43]
```diff
@@ -176,73 +172,114 @@ select.datetime {
...
+  line-height: 1.5rem;
+  input {
+    opacity: 0;
     position: absolute;
-    top: 1px;
-    left: 1px;
-    color: $linkColor;
+    top: 0;
+    left: 0;
+    &:focus + .control_styled {
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Selector `control_styled` should be written in lowercase with hyphens

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 44]
```diff
@@ -176,73 +172,114 @@ select.datetime {
...
+      border-color: $darkBlue;
+    }
+    &:hover + .control_styled {
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Selector `control_styled` should be written in lowercase with hyphens

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 45]
```diff
@@ -176,73 +172,114 @@ select.datetime {
...
+    &:active + .control_styled {
+      background: #f1f1f1;
+    }
+    &:checked + .control_styled:after {
+      opacity: 1;
+      transform: scale(1) translateZ(0);
+    }
+    &[readonly],
+    &[disabled] {
+      + .control_styled {
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Selector `control_styled` should be written in lowercase with hyphens

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 46]
```diff
@@ -176,73 +172,114 @@ select.datetime {
...
+        opacity: 0.6;
+        background: #f5f5f5;
+      }
+      &:active + .control_styled {
```
Comment (dobthound) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Selector `control_styled` should be written in lowercase with hyphens
